File: gitlfs-common/src/main/java/ru/bozaro/gitlfs/common/data/Operation.java
package ru.bozaro.gitlfs.common.data;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import javax.annotation.Nonnull;

/**
 * LFSP operation type.
 *
 * @author Artem V. Navrotskiy
 */
public enum Operation {
  Download {
    @Override
    public String toValue() {
      return "download";
    }

    @Override
    public <R> R visit(@Nonnull Visitor<R> visitor) {
      return visitor.visitDownload();
    }
  },
  Upload {
    @Override
    public String toValue() {
      return "upload";
    }

    @Override
    public <R> R visit(@Nonnull Visitor<R> visitor) {
      return visitor.visitUpload();
    }
  };

  @JsonCreator
  public static Operation forValue(@Nonnull String value) {
    for (Operation item : values()) {
      if (item.toValue().equals(value)) {
        return item;
      }
    }
    return null;
  }

  @JsonValue
  public abstract String toValue();

  public abstract <R> R visit(@Nonnull Visitor<R> visitor);

  public interface Visitor<R> {
    R visitDownload();

    R visitUpload();
  }
}

File: gitlfs-common/src/main/kotlin/ru/bozaro/gitlfs/common/data/Operation.kt
package ru.bozaro.gitlfs.common.data

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue

/**
 * LFSP operation type.
 *
 * @author Artem V. Navrotskiy
 */
enum class Operation {
    Download {
        override fun toValue(): String {
            return "download"
        }

        override fun <R> visit(visitor: Visitor<R>): R {
            return visitor.visitDownload()
        }
    },
    Upload {
        override fun toValue(): String {
            return "upload"
        }

        override fun <R> visit(visitor: Visitor<R>): R {
            return visitor.visitUpload()
        }
    };

    @JsonValue
    abstract fun toValue(): String
    abstract fun <R> visit(visitor: Visitor<R>): R
    interface Visitor<R> {
        fun visitDownload(): R
        fun visitUpload(): R
    }

    companion object {
        @JsonCreator
        fun forValue(value: String): Operation? {
            for (item in values()) {
                if (item.toValue() == value) {
                    return item
                }
            }
            return null
        }
    }
}
